File: DP/1289. Minimum Falling Path Sum II/1289. Minimum Falling Path Sum II.java
class Solution {
    // for every row to select the most minimu number and the second minimum number
    public int minFallingPathSum(int[][] arr) {
        // brute force dp
        // how to improve the inner loop to O(1)
        int len = arr.length;
        
        // get the first and second minimum element of initilized row
        int idxFirst = findMin(arr[0], -1);
        int idxSecond = findMin(arr[0], idxFirst);
        // for next follow row
        for (int i = 1; i < len; i++) {
            for (int j = 0; j < len; j++) {
                arr[i][j] += j == idxFirst ? arr[i-1][idxSecond] : arr[i-1][idxFirst];
            }
            idxFirst = findMin(arr[i], -1);
            idxSecond = findMin(arr[i], idxFirst);
        }
        
        // get the res on the last row;
        int res = Integer.MAX_VALUE;
        for (int i = 0; i < len; i++) {
            res = Math.min(res, arr[len-1][i]);
        }
        return res;
    }
    
    public int minFallingPathSumBruteForce(int[][] arr) {
        // brute force dp
        // how to improve the inner loop to O(1)
        int len = arr.length;
        int[][] dp = new int[len][len];
        // initilize
        for (int i = 0; i < len; i++) {
            dp[0][i] = arr[0][i];
        }
        // for next follow row
        for (int i = 1; i < len; i++) {
            for (int j = 0; j < len; j++) {
                boolean isFirst = false;
                for (int k = 0; k < len; k++) {
                    if (k != j) {
                        dp[i][j] = isFirst ?  Math.min(dp[i][j], dp[i-1][k] + arr[i][j]) : dp[i-1][k] + arr[i][j];
                        isFirst = true;
                    }
                }
            }
        }
        // get the res on the last row;
        int res = Integer.MAX_VALUE;
        for (int i = 0; i < len; i++) {
            res = Math.min(res, dp[len-1][i]);
        }
        return res;
    }
    
    private int findMin(int[] curRow, int startIdx) {
        int res = -1;
        for (int i = 0; i < curRow.length; i++) {
            if (startIdx == i) continue;
            if (res == -1) {
                res = i;
            } else if (curRow[res] > curRow[i]) {
                res = i;
            }
        }
        return res;
    }
}

File: DP/1289. Minimum Falling Path Sum II/1289. Minimum Falling Path Sum II.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& arr) {
        int m = arr.size();
        int n = arr[0].size();
        
        auto dp = vector<vector<int>>(m, vector<int>(n));
        // first step 
        for (int j = 0; j < n; j++) {
            dp[0][j] = arr[0][j];
        }
        
        for (int i = 1; i < m; i++) {
            // get the min1 and min2 of last row
            vector<pair<int, int>> temp;
            for (int k = 0; k < n; k++) {
                temp.push_back({dp[i-1][k], k});
            }
            sort(temp.begin(), temp.end());
            
            for (int j = 0; j < n; j++) {
                if (j != temp[0].second) {
                    dp[i][j] = temp[0].first +arr[i][j];
                } else {
                    dp[i][j] = temp[1].first +arr[i][j] ;
                }
                //dp[i][j] += arr[i][j];
            }
        }
        
        // last step 
        int ret = INT_MAX;
        for (int j = 0; j < n; j++) {
            ret = min(ret, dp[m-1][j]);
        }
        return ret;
    }
};

//dp[i][j] -> the minimum sum of a falling path with non-zero shifts from 0 row to {i, j}
//dp[i][j] = min/min2(dp[i][k] {k = 0,1,2,3,....n-1; k != j}) + arr[i][j]
